consensus/istanbul/backend/announce.go

```diff
@@ -700,3 +716,169 @@ func (sb *Backend) checkPeersAnnounceVersions() {
 		}
 	}
 }
+
+type versionedEnode struct {
+	Node    string
+	Version uint
+}
+
+// ==============================================
+//
+// define the functions that needs to be provided for rlp Encoder/Decoder.
+
+// EncodeRLP serializes ve into the Ethereum RLP format.
+func (ve *versionedEnode) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, []interface{}{ve.Node, ve.Version})
+}
+
+// DecodeRLP implements rlp.Decoder, and load the ve fields from a RLP stream.
+func (ve *versionedEnode) DecodeRLP(s *rlp.Stream) error {
+	var msg struct {
+		Node    string
+		Version uint
+	}
+
+	if err := s.Decode(&msg); err != nil {
+		return err
+	}
+	ve.Node, ve.Version = msg.Node, msg.Version
+	return nil
+}
+
+// retrieveSelfVersionedEnodeMsg gets the most recent self versioned enode message to send
+// and generates a new one with the current announce version if one does not exist.
+// New versioned enode messages are not always generated to ensure the version
+// of a versioned enode message is not greater than a recently gossiped announce
+// version (if that has occurred)
+// May be nil if this is a proxy that does not have a versioned enode message
+// from its proxy.
+func (sb *Backend) retrieveSelfVersionedEnodeMsg() (*istanbul.Message, error) {
+	sb.selfVersionedEnodeMsgMu.Lock()
+	defer sb.selfVersionedEnodeMsgMu.Unlock()
+	if sb.selfVersionedEnodeMsg == nil {
+		// Proxies cannot generate a versioned enode message because are not
+		// able to sign on behalf of the proxied validator.
+		if sb.config.Proxy {
+			return nil, nil
+		}
+		versionedEnodeMsg, err := sb.generateVersionedEnodeMsg(getCurrentAnnounceVersion())
```

Comment:
When the node first starts up there's at least a 1 minute period before the first call to `generateAndGossipAnnounce` (that's how I'm reading the code, correct me if I'm wrong)
I think that first minute is likely to be a time period where this handshake would be very useful (ie new validator connections from a new node)
I tried adding this to `announceThread` when it's first called, but by that time the istanbul signing function is not set yet.

Reply:
My main concern here is that the enode certificate and version number should only be generated when the enodeURL assignment is changed.

As it's currently implemented, those would be updated and generated on every call to `retrieveSelfVersionedEnodeMsg`.  That results in two issues:

1) The certificate will be generated every time `registeredpeer` is called, so there is going to be alot of redundant work done by the done.
2) More importantly, the version number will be updated every time `registeredpeer` is called, and hence there will be an expensive gossiped message every time that function is called.  That will also potentially make all the other validators gossip queryMessage (another *very* expensive networking call).

Reply:
`retrieveVersionedEnodeMsg` will only generate the enode certificate & version number if one hasn't been previously generated, which can only happen once upon the first call to `retrieveVersionedEnodeMsg`. Otherwise the previously generated enode certificate & version number is returned

So the only way the enode certificate/version is updated is if:

1. `generateAndGossipAnnounce` generates a new announce message
2. The node is a validator and the enode certificate hasn't been previously generated by `generateAndGossipAnnounce` and the certificate is needed for the handshake or in `RegisterPeer` if the peer is a proxy
3. The proxy receives an enode certificate from the proxied validator as a result of a proxied validator's newly generated announce or the proxied validator's `RegisterPeer`